pkg/kube/util/ctxlog/events.go

```diff
@@ -43,15 +47,49 @@ func WithEvent(object runtime.Object, reason string) EventLogger {
 	return event{object: object, reason: reason}
 }
 
+func (ev event) DebugPredicate(ctx context.Context, reason string, meta metav1.Object, t string, msg string) {
+	ev.DebugJSON(
+		ctx,
+		reason,
+		ReconcileEventsFromSource{
+			ReconciliationObjectName: meta.GetName(),
+			ReconciliationObjectKind: t,
+			PredicateObjectName:      meta.GetName(),
+			PredicateObjectKind:      t,
+			Namespace:                meta.GetNamespace(),
+			Type:                     "predicate",
+			Message:                  msg,
+		},
+	)
+}
+
+func (ev event) DebugMapping(ctx context.Context, reconciliation reconcile.Request, crd string, objName string, objType string) {
+	ev.DebugJSON(
+		ctx,
+		"Enqueuing reconcile requests in response to events",
+		ReconcileEventsFromSource{
+			ReconciliationObjectName: reconciliation.Name,
+			ReconciliationObjectKind: crd,
+			PredicateObjectName:      objName,
+			PredicateObjectKind:      objType,
+			Namespace:                reconciliation.Namespace,
+			Type:                     "mapping",
+			Message:                  fmt.Sprintf("fan-out updates from %s, type %s into %s", objName, objType, reconciliation.Name),
+		},
+	)
+}
+
 // DebugJSON logs and adds an info event in json format
-func (ev event) DebugJSON(ctx context.Context, format string, objectInfo interface{}) {
+func (ev event) DebugJSON(ctx context.Context, msg string, objectInfo interface{}) {
 	log := ExtractLogger(ctx)
 
 	jsonData, _ := json.Marshal(objectInfo)
-	log.Debug(format, string(jsonData))
+	log.Debug(string(jsonData))
 
 	recorder := ExtractRecorder(ctx)
-	recorder.Event(ev.object, corev1.EventTypeNormal, ev.reason, fmt.Sprintf("%s %s", format, string(jsonData)))
+	var result map[string]string
+	json.Unmarshal([]byte(jsonData), &result)
+	recorder.Event(ev.object, corev1.EventTypeNormal, ev.reason, fmt.Sprintf("%s %s", msg, result["message"]))
```

Comment:
@manno  from your comment in the PR
> I think the JSON should be in events, not in the debug log. Events only live for about an hour and are machine readable via the k8s AP

Why we are you not doing this now? I have the same opinion as you, keep the JSON in the events. Also, if the JSON will come here in events, the `msg` is not needed, either embedded inside the JSON, or remove it.

The rest of the code changes is nice, more compact, thanks!